File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/news/details/ArticleContentFragment.java
package forpdateam.ru.forpda.ui.fragments.news.details;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.JavascriptInterface;

import forpdateam.ru.forpda.api.news.models.DetailsPage;
import forpdateam.ru.forpda.apirx.RxApi;
import forpdateam.ru.forpda.common.webview.CustomWebChromeClient;
import forpdateam.ru.forpda.common.webview.CustomWebViewClient;
import forpdateam.ru.forpda.ui.activities.MainActivity;
import forpdateam.ru.forpda.ui.views.ExtendedWebView;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by radiationx on 03.09.17.
 */

public class ArticleContentFragment extends Fragment {
    public final static String JS_INTERFACE = "INews";
    private ExtendedWebView webView;
    private DetailsPage article;

    public ArticleContentFragment setArticle(DetailsPage article) {
        this.article = article;
        return this;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        webView = ((MainActivity) getActivity()).getWebViewsProvider().pull(getContext());
        registerForContextMenu(webView);
        webView.setWebViewClient(new CustomWebViewClient());
        webView.setWebChromeClient(new CustomWebChromeClient());
        webView.addJavascriptInterface(this, JS_INTERFACE);
        loadHtml();
        return webView;
    }

    private void loadHtml() {
        webView.loadDataWithBaseURL("https://4pda.ru/forum/", article.getHtml(), "text/html", "utf-8", null);
    }

    @JavascriptInterface
    public void toComments() {
        if (getContext() == null)
            return;
        webView.runInUiThread(() -> {
            ((NewsDetailsFragment) getParentFragment()).getFragmentsPager().setCurrentItem(1);
        });
    }

    @JavascriptInterface
    public void sendPoll(String id, String answer, String from) {
        if (getContext() == null)
            return;
        webView.runInUiThread(() -> {
            int pollId = Integer.parseInt(id);
            String[] answers = answer.split(",");
            int answersId[] = new int[answers.length];
            for (int i = 0; i < answers.length; i++) {
                answersId[i] = Integer.parseInt(answers[i]);
            }
            NewsDetailsFragment fragment = ((NewsDetailsFragment) getParentFragment());
            fragment.subscribe(RxApi.NewsList().sendPoll(from, pollId, answersId), page -> {
                article.setHtml(page.getHtml());
                loadHtml();
            }, new DetailsPage());
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        if (webView != null) {
            webView.onResume();
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (webView != null) {
            webView.onPause();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (webView != null) {
            webView.endWork();
            ((MainActivity) getActivity()).getWebViewsProvider().push(webView);
        }
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/news/details/ArticleContentFragment.kt
package forpdateam.ru.forpda.ui.fragments.news.details

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.JavascriptInterface
import androidx.lifecycle.lifecycleScope
import forpdateam.ru.forpda.common.Utils
import forpdateam.ru.forpda.common.webview.CustomWebChromeClient
import forpdateam.ru.forpda.common.webview.CustomWebViewClient
import forpdateam.ru.forpda.common.webview.DialogsHelper
import forpdateam.ru.forpda.entity.remote.news.DetailsPage
import forpdateam.ru.forpda.extensions.quillMoxyPresenter
import forpdateam.ru.forpda.presentation.LinkHandler
import forpdateam.ru.forpda.presentation.SystemLinkHandler
import forpdateam.ru.forpda.presentation.TabRouter
import forpdateam.ru.forpda.presentation.articles.detail.content.ArticleContentPresenter
import forpdateam.ru.forpda.presentation.articles.detail.content.ArticleContentView
import forpdateam.ru.forpda.ui.fragments.TabTopScroller
import forpdateam.ru.forpda.ui.fragments.WebViewTopScroller
import forpdateam.ru.forpda.ui.views.ExtendedWebView
import kotlinx.coroutines.launch
import moxy.MvpAppCompatFragment
import ru.radiationx.coretypes.ArticleAnswerId
import ru.radiationx.coretypes.ArticlePollId
import ru.radiationx.quill.inject

/**
 * Created by radiationx on 03.09.17.
 */

class ArticleContentFragment : MvpAppCompatFragment(), ArticleContentView, TabTopScroller {

    private val webView: ExtendedWebView
        get() = requireView() as ExtendedWebView

    private lateinit var topScroller: WebViewTopScroller

    private val utils by inject<Utils>()
    private val linkHandler by inject<LinkHandler>()
    private val systemLinkHandler by inject<SystemLinkHandler>()
    private val router by inject<TabRouter>()
    private val webViewClient by inject<CustomWebViewClient>()
    private val presenter by quillMoxyPresenter<ArticleContentPresenter>()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return ExtendedWebView(requireContext())
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        (parentFragment as? NewsDetailsFragment)?.attachWebView(webView)
        topScroller =
            WebViewTopScroller(webView, (parentFragment as NewsDetailsFragment).getAppBar())
        webView.setDialogsHelper(
            DialogsHelper(
                context = webView.context,
                linkHandler = linkHandler,
                systemLinkHandler = systemLinkHandler,
                utils = utils,
                router = router,
            )
        )
        registerForContextMenu(webView)
        webView.webViewClient = webViewClient
        webView.webChromeClient = CustomWebChromeClient()
        webView.addJavascriptInterface(this, JS_INTERFACE)
    }

    override fun toggleScrollTop() {
        topScroller.toggleScrollTop()
    }

    override fun setRefreshing(isRefreshing: Boolean) {}

    override fun showData(article: DetailsPage) {
        webView.loadDataWithBaseURL(
            "https://4pda.to/forum/",
            article.html,
            "text/html",
            "utf-8",
            null
        )
    }

    override fun setStyleType(type: String) {
        webView.evalJs("changeStyleType(\"$type\")")
    }

    override fun setFontSize(size: Int) {
        webView.setRelativeFontSize(size)
    }

    @JavascriptInterface
    fun toComments() {
        viewLifecycleOwner.lifecycleScope.launch {
            (parentFragment as NewsDetailsFragment).fragmentsPager.currentItem = 1
        }
    }

    @JavascriptInterface
    fun sendPoll(id: String, answer: String, from: String) {
        viewLifecycleOwner.lifecycleScope.launch {
            val pollId = ArticlePollId(id.toInt())
            val answerIds = answer.split(",".toRegex())
                .dropLastWhile { it.isEmpty() }
                .map { ArticleAnswerId(it.toInt()) }
            presenter.sendPoll(from, pollId, answerIds)
        }
    }

    override fun onResume() {
        super.onResume()
        webView.onResume()
    }

    override fun onPause() {
        super.onPause()
        webView.onPause()
    }

    override fun onDestroy() {
        super.onDestroy()
        webView.endWork()
    }

    companion object {
        const val JS_INTERFACE = "INews"
    }
}
